### PR TITLE
Add ResourceFilesTab and corresponding views

### PR DESCRIPTION
**Pull request by gagelarsen**
**Primary changes in this Pull Request:**

- Add `ResourceFilesTab` class and resulting views.R

**Please review the checklist before submitting the Pull Request:**

- [x] Pull request has a meaning full name (not just the commit message from of your last commit)
- [x] Code has been linted using flake8
- [x] All methods have accurate Google-Style Docstrings
- [x] 100% test coverage for new content
- [x] Tests for each common use case (please don't write one test that covers all use cases)
- [x] **kinda...** Bonus: Use TypeHints

**Explain deviations from original design if applicable:**

- Add expand/collapse folders
- Sortable Table

### CODE REVIEW

File: tethysext/atcore/tests/integrated_tests/controllers/resources/tabs/files_tab_tests.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        """Test that "All Users" is displayed in debug information when resource is locked for all users."""
+        instance = ResourceFilesTab()
+        request = self.request_factory.get('/foo/12345/bar/summary/')
+        user = self.get_user(is_staff=True)
+        request.user = user
+        self.resource.acquire_user_lock()
+
+        context = instance.get_context(request, self.session, self.resource, {})
+
+        self.assertIn('collections', context)
```
Comment (swainn) [inline tethysext/atcore/tests/integrated_tests/controllers/resources/tabs/files_tab_tests.py:117]: Do these three tests apply? Does the type of user matter to the ResourceFilesView?

**Review by swainn — COMMENTED** [review]
Need to see a demo and get Download implemented. See other comments too.

**Comment by swainn** [comment]
Two other features: 

- [ ] exand all/collapse all

- [ ] sort viewport table by columns (use datatable?)

File: tethysext/atcore/controllers/app_users/manage_resources.py [hunk 2]
```diff
@@ -286,7 +286,8 @@ def get_resources(self, session, request, request_app_user):
         Returns:
             list<Resources>: the list of resources to render on the manage_resources page.
         """
-        return request_app_user.get_resources(session, request)
+        of_type = self.get_resource_model()
+        return request_app_user.get_resources(session, request, of_type=of_type)
```
[the remark this answers is not in the document]
Comment (gagelarsen) [reply tethysext/atcore/controllers/app_users/manage_resources.py:290]: I did a rebase

File: tethysext/atcore/tests/integrated_tests/controllers/app_users/manage_resources.py [hunk 3]
```diff
@@ -402,17 +402,18 @@ def test_get_resource_action_launch(self, mock_reverse):
...
             }, ret)
 
-    def test_get_resources(self):
+    @mock.patch('tethysext.atcore.controllers.app_users.mixins.AppUsersViewMixin.get_resource_model')
+    def test_get_resources(self, mock_get_resource_model):
         mock_request = self.request_factory.get('/foo/bar/')
         mock_session = mock.MagicMock()
         mock_request_app_user = mock.MagicMock()
-
+        mock_get_resource_model.return_value = 'resource_type'
```
[the remark this answers is not in the document]
Comment (gagelarsen) [reply tethysext/atcore/tests/integrated_tests/controllers/app_users/manage_resources.py:410]: I was wondering that too.

File: tethysext/atcore/controllers/resources/tabs/files_tab.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+            # Calculate the file size and convert to the appropriate measurement.
+            power = 2 ** 10
+            n = 0
+            power_labels = {0: 'Bytes', 1: 'KB', 2: 'MB', 3: 'GB', 4: 'TB'}
+            size = os.path.getsize(path)
+            while size > power:
+                size /= power
+                n += 1
+            size_str = f'{size:.1f}' if size > 0 else '0'
+            hierarchy['size'] = f'{size_str} {power_labels[n]}'
```
[the remark this answers is not in the document]
Comment (gagelarsen) [reply tethysext/atcore/controllers/resources/tabs/files_tab.py:111]: It was, but not as fun as building the file hierarchy stuff...

File: tethysext/atcore/controllers/resources/tabs/files_tab.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+import errno
+import os
+import time
+
+from .resource_tab import ResourceTab
+
+
+class ResourceFilesTab(ResourceTab):
+    """
+    A tab for the TabbedResourceDetails view that lists key-value pair attributes of the Resource. The attributes can be grouped into multiple sections with titles.
```
[the remark this answers is not in the document]
Comment (gagelarsen) [reply tethysext/atcore/controllers/resources/tabs/files_tab.py:18]: Fixed.

File: tethysext/atcore/controllers/resources/tabs/files_tab.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    js_requirements = ResourceTab.js_requirements + [
+        'atcore/resources/files_tab.js'
+    ]
+    css_requirements = ResourceTab.css_requirements + [
+        'atcore/resources/files_tab.css'
+    ]
+
+    def get_file_collections(self, request, resource, *args, **kwargs):
+        """
+        Get the summary tab info
```
[the remark this answers is not in the document]
Comment (gagelarsen) [reply tethysext/atcore/controllers/resources/tabs/files_tab.py:39]: Fixed

File: tethysext/atcore/controllers/resources/tabs/files_tab.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+        Return Format
+        [FileCollectionClient, FileCollectionClient, FileCollectionClient]
```
[the remark this answers is not in the document]
Comment (gagelarsen) [reply tethysext/atcore/controllers/resources/tabs/files_tab.py:42]: Fixed

File: tethysext/atcore/controllers/resources/tabs/files_tab.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    post_load_callback = 'files_tab_loaded'
+
+    js_requirements = ResourceTab.js_requirements + [
+        'atcore/resources/files_tab.js'
+    ]
+    css_requirements = ResourceTab.css_requirements + [
+        'atcore/resources/files_tab.css'
+    ]
+
+    def get_file_collections(self, request, resource, *args, **kwargs):
```
[the remark this answers is not in the document]
Comment (gagelarsen) [reply tethysext/atcore/controllers/resources/tabs/files_tab.py:37]: Done

File: tethysext/atcore/public/resources/files_tab.js [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+                if (row.dataset.slug == elem.dataset.slug) {
+                    change_tree_selection(elem);
+                }
+            });
+        });
+    });
+
+    document.querySelectorAll('.viewport_file').forEach(function(row) {
+        $(row).bind("dblclick", function(){
+            alert("Download file: " + row.dataset.filepath);
```
[the remark this answers is not in the document]
Comment (gagelarsen) [reply tethysext/atcore/public/resources/files_tab.js:101]: Lets get together for this

File: tethysext/atcore/templates/atcore/resources/tabs/files_tab.html [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+            {# Loop through and add a table row for each item in the collection, we will handle
+               hiding and how the items in the javascript. #}
+            {% for collection_id, collection_contents in collections.items %}
+                {%for node in collection_contents.children %}
+                    {%include "atcore/resources/tabs/files_viewport_template.html" %}
+                {%endfor%}
+            {% endfor %}
+            </tbody>
+        </table>
+    </div>
```
[the remark this answers is not in the document]
Comment (gagelarsen) [reply tethysext/atcore/templates/atcore/resources/tabs/files_tab.html:55]: Sorry

File: tethysext/atcore/tests/integrated_tests/controllers/resources/tabs/files_tab_tests.py (hunk 1 above)
Comment (gagelarsen) [reply tethysext/atcore/tests/integrated_tests/controllers/resources/tabs/files_tab_tests.py:117]: Not really, i removed them

File: tethysext/atcore/public/resources/files_tab.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (gagelarsen) [reply tethysext/atcore/public/resources/files_tab.js]: We did this.

File: tethysext/atcore/controllers/resources/tabs/files_tab.py [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+            'parent_path': os.path.abspath(os.path.join(hierarchy_path, os.pardir)).replace(root_dir, ''),
+            'parent_slug': parent_slug,
+            'slug': '_' + hierarchy_path.replace(os.path.sep, '_').replace('.', '_').replace('-', '_'),
+        }
+
+        # Try and access 'children' here. If we can't than this is a file.
+        try:
+            # Recurse through each of the children if it is a directory.
+            hierarchy['children'] = [
+                self._path_hierarchy(os.path.join(path, contents), root_dir, hierarchy['slug'])
```
[the remark this answers is not in the document]
Comment (gagelarsen) [reply tethysext/atcore/controllers/resources/tabs/files_tab.py]: unless you prefer list comprehensions. Done.

File: tethysext/atcore/controllers/resources/tabs/files_tab.py [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+        Return Format
+        [FileCollectionClient, FileCollectionClient, FileCollectionClient]
+        """
+        return []
+
+    def get_context(self, request, session, resource, context, *args, **kwargs):
+        """
+        Build context for the ResourceFilesTab template that is used to generate the tab content.
+        """
+        collections = self.get_file_collections(request, resource)
```
[the remark this answers is not in the document]
Comment (gagelarsen) [reply tethysext/atcore/controllers/resources/tabs/files_tab.py]: done

File: tethysext/atcore/controllers/resources/tabs/files_tab.py [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+        """
+        A function used to create a dictionary representation of a folder structure.
+
+        Args:
+            path: The path to recursively map to a dictionary.
+            root_dir: The root directory to be trimmed off of the absolute paths.
+            parent_slug: The slug for the parent used for hiding and showing files.
+
+        Returns:
+            A dictionary defining the folder structure of the provided path.
```
[the remark this answers is not in the document]
Comment (gagelarsen) [reply tethysext/atcore/controllers/resources/tabs/files_tab.py]: done

**Review by swainn — DISMISSED** [review]
This looks good. Just need to fix the git history.

**Review by swainn — CHANGES_REQUESTED** [review]
Need to address git history issue.

**Review by swainn — APPROVED** [review]

**Pull request merged**